src/compile/data/calculate.ts

```diff
@@ -53,24 +51,64 @@ export class CalculateNode extends DataFlowNode {
     return parent;
   }
 
+  /**
+   * Merge calculate nodes. This method either integrates the bin config from the other node
+   * or if this node already has a bin config, renames the corresponding signal in the model.
+   */
+  public merge(other: CalculateNode) {
+    for (const key of keys(other.calculates)) {
+      if (key in this.calculates) {
+        // make sure we are not merging something incompatible
+        if (this.calculates[key].calculate !== other.calculates[key].calculate) {
+          /* istanbul ignore next: This should never happen. */
+          throw new Error('Merged incompatible calculates.');
+        }
+      } else {
+        this.calculates[key] = other.calculates[key];
+      }
+    }
+
+    for (const child of other.children) {
+      other.removeChild(child);
+      child.parent = this;
+    }
+    other.remove();
+  }
+
   public producedFields() {
-    return new Set([this.transform.as]);
+    return new Set(vals(this.calculates).map(c => c.as));
   }
 
   public dependentFields() {
-    return this._dependentFields;
+    let depFields = new Set<string>();
+
+    for (const c of vals(this.calculates)) {
+      depFields = new Set([...depFields, ...getDependentFields(c.calculate)]);
+    }
+
+    return depFields;
   }
 
-  public assemble(): VgFormulaTransform {
-    return {
-      type: 'formula',
-      expr: this.transform.calculate,
-      as: this.transform.as
-    };
+  public assemble(): VgFormulaTransform[] {
+    const transforms: VgFormulaTransform[] = [];
+
+    for (const {key, value} of entries(this.calculates)) {
+      if (key !== value.as) {
+        /* istanbul ignore next: This should never happen. */
```

Comment:
Why checking this if it shouldn't ever happen?  

(If you worry about bug, can you check this in a unit test instead? )

Reply:
This is an assertion that I added. It won't happen right now but I am worried that at some point I might forget to satisfy the assumptions.